control-plane/cni/main.go

```diff
@@ -135,56 +162,120 @@ func cmdAdd(args *skel.CmdArgs) error {
-		annotations := map[string]string{
-			keyCNIStatus: "true",
-		}
-		pod.SetAnnotations(annotations)
-		_, err = client.CoreV1().Pods(podNamespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
-		if err != nil {
-			return err
-		}
-	} else {
-		logger.Debug("skipping traffic redirect on un-injected pod")
+	// skip injecting pod that has consul.hashicorp.com/connect-inject: false
+	if skipConnectInject(*pod) {
+		logger.Debug("skipping traffic redirect on un-injected pod: %s", pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// check to see if the cni-proxy-config annotation exists and if not, wait, retry and backoff.
+	exists := waitForAnnotation(*pod, annotationCNIProxyConfig, uint64(retries))
+	if !exists {
+		logger.Error("could not retrieve annotation: %s on pod: %s", annotationCNIProxyConfig, pod.Name)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// parse the cni-proxy-config annotation into an iptables.Config object.
+	iptablesCfg, err := parseAnnotation(*pod, annotationCNIProxyConfig)
+	if err != nil {
+		logger.Error("could not parse annotation: %s, error: %v", annotationCNIProxyConfig, err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// get the DNS IP from the environment.
+	dnsIP := searchDNSIPFromEnvironment(*pod, cfg.DNSPrefix)
+	if dnsIP != "" {
+		logger.Error("assigned consul dns ip to %s", dnsIP)
+		iptablesCfg.ConsulDNSIP = dnsIP
 	}
 
-	// TODO: Get transparent proxy annotations and merge with proxy config
-	// TODO: Redirect traffic :)
+	// apply the iptables rules.
+	err = iptables.Setup(iptablesCfg)
+	if err != nil {
+		logger.Error("could not apply iptables setup: %v", err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
 
-	// Pass through the result for the next plugin
+	logger.Debug("traffic redirect rules applied to pod: %s", pod.Name)
+	// Pass through the result for the next plugin even though we are the final plugin in the chain.
 	return types.PrintResult(result, cfg.CNIVersion)
 }
 
 // cmdDel is called for DELETE requests.
 func cmdDel(args *skel.CmdArgs) error {
-	// Nothing for consul-cni plugin to do as everything is removed once the pod is gone
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
+// cmdCheck is called for CHECK requests.
 func cmdCheck(args *skel.CmdArgs) error {
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
 func main() {
 	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.All, bv.BuildString("consul-cni"))
 }
 
-func hasBeenInjected(pod corev1.Pod) bool {
-	if anno, ok := pod.Annotations[keyInjectStatus]; ok && anno == injected {
+// skipConnectInject determines if the connect-inject annotation is false.
+func skipConnectInject(pod corev1.Pod) bool {
+	if anno, ok := pod.Annotations[annotationInject]; ok && anno == "false" {
 		return true
 	}
 	return false
 }
```

Comment:
What about the behavior with the `connectInject.default` Helm value? If you have it set to false, then you wouldn't need to set connect inject annotation to false (because it's redundant). This function will return false, but it should return true. 

We should probably instead look at connect inject status like we do in the `shouldInject` function on the webhook which also checks for other rules.

Reply:
Still working on this piece.

Reply:
This is addressed in a further PR when we do traffic redirect and the webhook work